[c10d] user to explicitly specify whether split semantics should be used in new_group

shuqiangzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138518

Summary:
Eager init and split should be two different concepts. E.g., we should be able to support eager init but without using split mode. 

In this PR, we provide users an option to specify use_split or not, even if eager init is used.

Also, in the future, it is recommended to use split_group instead of new_group if users want to use split. 

Test Plan:
Added a test_comm_eager_without_split
Reviewers:

Subscribers:

Tasks:

Tags:

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

shuqiangzhang committed:
[c10d] explicity specify if split semantics should be used
Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d] user to explicitly specify if split semantics should be used in new_group"


Summary:
Eager init and split should be two different concepts. E.g., we should be able to support eager init but without using split mode. 

In this PR, one needs to explicitly specify use_split, in order to use the split comm functions in NCCL

Test Plan:
Added a test_comm_eager_without_split
Reviewers:

Subscribers:

Tasks:

Tags:

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d] user to explicitly specify if split semantics should be used in new_group"


Summary:
Eager init and split should be two different concepts. E.g., we should be able to support eager init but without using split mode. 

In this PR, one needs to explicitly specify use_split, in order to use the split comm functions in NCCL

Test Plan:
Added a test_comm_eager_without_split
Reviewers:

Subscribers:

Tasks:

Tags:

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d] user to explicitly specify if split semantics should be used in new_group"


Summary:
Eager init and split should be two different concepts. E.g., we should be able to support eager init but without using split mode. 

In this PR, we provide users an option to specify use_split or not, even if eager init is used.

Test Plan:
Added a test_comm_eager_without_split
Reviewers:

Subscribers:

Tasks:

Tags:

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented:
It depends on whether we see split as implementation detail or not.

shuqiangzhang commented:
> It depends on whether we see split as implementation detail or not.

This is similar to eager init/device id para input. For normal users where init time or resource does not matter much, they might not need to know or set this. But if they care, they have a choice

wconstab commented:
my thought is that use_split defaults to false, which may lead to less usage of split.  I think we need to balance between good design to separate split/eager, but also using defaults to encourage the path we want to encourage.

What are the tradeoffs of using new_group + use_split=True vs False?  What about the tradeoffs of using new_group+use_split vs split_group?

Separately, if you want to discourage new_group and encourage split_group, how do we do that?

shuqiangzhang commented:
> my thought is that use_split defaults to false, which may lead to less usage of split. I think we need to balance between good design to separate split/eager, but also using defaults to encourage the path we want to encourage.
> 
> What are the tradeoffs of using new_group + use_split=True vs False? What about the tradeoffs of using new_group+use_split vs split_group?
> 
> Separately, if you want to discourage new_group and encourage split_group, how do we do that?

Right, Checking my commit V2, V3 and V4, you could see me debating myself by changing the default use_split from true to false and then to true again. 

The intention of changing the default to be false is exactly to encourage performance caring users to use split_group instead of new_group. split_group is mostly stable now and is better than new_group + use_split=True for all 2 strong reasons listed in https://github.com/pytorch/pytorch/issues/130407?fbclid=IwZXh0bgNhZW0CMTEAAR1rsWYer3sssb73mX5fkMV_JeiN8sy-cK55YrFoleVKbJNQX031LWb-dPI_aem_KMgnegupSqFGoVhs4Sblbg, that's why advanced users (including device mesh) are using split_group now. 

Of course, we can only suggest this to users (modified comment/doc) but not force it, that's why we still need to keep the old option viable of new_group. For users don't care about perf, not using split is fine. But once users feel the needs of better performance, they can follow the suggestion

kwen2501 commented:
Shall we look at the `new_group` API solely?

There are a couple questions we'd need to answer imo:

- For `new_group`, is split-based impl strictly better than bootstrap-based impl?
- For `new_group`, do users need to know / express which impl is used?

shuqiangzhang commented:
> Shall we look at the `new_group` API solely?
> 
> There are a couple questions we'd need to answer imo:
> 
> * For `new_group`, is split-based impl strictly better than bootstrap-based impl?
> * For `new_group`, do users need to know / express which impl is used?

From perf point of view,  newgroup without split < newgroup with split < split_group. For normal OSS users, they don't need to know the details, that's why it's an optional para.  Do you suggest we make split to be default? I am totally OK with that as that is the originally plan.

shuqiangzhang committed:
Update on "[c10d] user to explicitly specify whether split semantics should be used in new_group"


Summary:
Eager init and split should be two different concepts. E.g., we should be able to support eager init but without using split mode. 

In this PR, we provide users an option to specify use_split or not, even if eager init is used.

Also, in the future, it is recommended to use split_group instead of new_group if users want to use split. 

Test Plan:
Added a test_comm_eager_without_split
Reviewers:

Subscribers:

Tasks:

Tags:

cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

wconstab commented:
One more clarification- if you use new group with or without split, you still have to have all ranks call new group the same number of times in the same order, right? (No visible semantic changes based on the split flag)?

shuqiangzhang commented:
> One more clarification- if you use new group with or without split, you still have to have all ranks call new group the same number of times in the same order, right? (No visible semantic changes based on the split flag)?
That's a good question that I hadn't thought before. Indeed, if no split is used, I think there is no need of the requirement of calling new_group on every rank of the default PG, so this requirement can be relaxed. Though calling it on every ranks does not do any harm and just backward compatible
